src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java

```diff
@@ -477,6 +477,33 @@ private void trustedShell(final boolean pos) {
         });
     }
 
+    @Test
+    public void existingVariablesOnRestart() throws Exception {
+        story.then(r -> {
+            WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+            p.setDefinition(new CpsFlowDefinition("echo('pre-semaphore')\n" +
+                    "semaphore('wait')\n" +
+                    "echo('post-semaphore')", true));
+            WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+            SemaphoreStep.waitForStart("wait/1", b);
+        });
+        story.then(r -> {
+            WorkflowJob p = r.jenkins.getItemByFullName("p", WorkflowJob.class);
+            WorkflowRun b = p.getBuildByNumber(1);
+            SemaphoreStep.success("wait/1", null);
+            r.waitForCompletion(b);
+            r.assertBuildStatus(Result.SUCCESS, b);
+        });
+    }
+
+    @TestExtension("ExistingVariablesOnRestart")
```

Comment:
```suggestion
    @TestExtension("existingVariablesOnRestart")
```

The fact that the test passes despite this mistake is a bad sign. Did you verify the bug symptom with `src/main/` changes reverted? The test is not asserting anything about this particular variable, only that the build does not die.

Reply:
> manually checked that the injected variables are present before and after the restart by dumping the program state as XML

Interesting, but should be easy enough to just `echo` the variable value in an automated test as well.

Reply:
addressed in f15b02ddf3268eac1df80d25f14745975d68baa4